### PR TITLE
Bug fix in QIO buffering optimization

### PR DESCRIPTION
**Pull request by jeremiah-corrado**
Fixes a bug in the IO runtime where a `fileReader` could read from the incorrect file offset when the runtime transitioned from buffered reading to non-buffered reading.

The bug was somewhat unlikely to come up in practice, as the total size of the buffered reads needed to add up to exactly the size of one qio buffer chunk (2**16 bytes).

Resolves: https://github.com/chapel-lang/chapel/issues/23800

- [x] local paratest
- [x] gasnet paratest
- [x] test referenced in [#23800](https://github.com/chapel-lang/chapel/issues/23800) is passing

### CODE REVIEW

File: runtime/src/qio/qio.c [hunk 1]
```diff
@@ -2829,26 +2829,34 @@ qioerr _qio_buffered_read(qio_channel_t* ch, void* ptr, ssize_t len, ssize_t* am
...
+    *amt_read = gotlen;
 
-      // clean up the now unused portion of the buffer
-      qbuffer_trim_front(&ch->buf, gotlen);
-    }
+    // clean up the now unused portion of the buffer
+    qbuffer_trim_front(&ch->buf, gotlen);
+    ch->cached_cur = NULL;
+    ch->cached_end = NULL;
+    ch->cached_start = NULL;
```
Comment (mppf) [inline runtime/src/qio/qio.c:2859]: Why is `cache_cur` set on line 2851 only to be cleared out here?

File: runtime/src/qio/qio.c [hunk 2]
```diff
@@ -2919,6 +2926,7 @@ qioerr _qio_buffered_read(qio_channel_t* ch, void* ptr, ssize_t len, ssize_t* am
 
       // figure out the end of the data to copy
       gotlen = ch->av_end - _right_mark_start(ch);
+      // printf("\t\tfull gotlen: %lli\n", gotlen);
```
Comment (mppf) [inline runtime/src/qio/qio.c:2929]: Probably this needs to be removed

File: runtime/src/qio/qio.c [hunk 3]
```diff
@@ -2883,7 +2891,6 @@ qioerr _qio_buffered_read(qio_channel_t* ch, void* ptr, ssize_t len, ssize_t* am
       }
       // Return early on an error or on EOF.
       if( err ) {
-        *amt_read = len - remaining;
```
Comment (mppf) [inline runtime/src/qio/qio.c:2886]: Why is this line removed?

**Review by mppf — APPROVED** [review]

File: runtime/src/qio/qio.c (hunk 1 above)
Comment (jeremiah-corrado) [reply runtime/src/qio/qio.c:2859]: Setting `cache_cur` there is unneeded. I've removed it.

File: runtime/src/qio/qio.c (hunk 3 above)
Comment (jeremiah-corrado) [reply runtime/src/qio/qio.c:2886]: This was a second minor bug that was invalidating `amt_read` when the file pointer is at EOF.

File: runtime/src/qio/qio.c [hunk 4]
```diff
@@ -2890,6 +2890,7 @@ qioerr _qio_buffered_read(qio_channel_t* ch, void* ptr, ssize_t len, ssize_t* am
       }
       // Return early on an error or on EOF.
       if( err ) {
+        // do not update 'amt_read' because zero bytes were read on EOF
```
Comment (mppf) [inline runtime/src/qio/qio.c:2893]: a nit: amt_read was already update in the above `case` statements; it's just that we don't need to act on it.

Comment (jeremiah-corrado) [reply runtime/src/qio/qio.c:2893]: do you mean `num_read` was already updated in the case statements? (`amt_read` was updated in the previous iteration of the while loop)

Comment (mppf) [reply runtime/src/qio/qio.c:2893]: Oh, sorry, yes, you are right. Ignore my comment please!

**Comment by bradcray** [comment]
@jeremiah-corrado / @mppf: I'm feeling concerned about this fixing a bug but not adding a test to lock in the fix.  Arguably, once the future from #23800 is retired, that will serve as one such test, but Jeremiah's comment:

> The bug was somewhat unlikely to come up in practice, as the total size of the buffered reads needed to add up to exactly the size of one qio buffer chunk (2**16 bytes).

makes me worry that if someone were to change the qio buffer chunk in the future, that test would continue working if it didn't happen to use the new chunk size.  Is the chunk size something that's available from Chapel code, and if so, would it be possible to add a test that is based on that chunk size so that if/when it changes again, the test will change with it?

**Comment by jeremiah-corrado** [comment]
> Is the chunk size something that's available from Chapel code

```chapel
extern var qbytes_iobuf_size: c_size_t;
```

There is nothing user facing, but I'll modify the [growing_buff_read_write.chpl](https://github.com/chapel-lang/chapel/pull/23805/files#diff-4eec6282a07b9a84180ac5ee5412d08d9057672704b42c9dac9354f3527bf0a4) s.t. reads will add up to that size.
